core/modules/loader/BufferUdp.cc

```diff
@@ -0,0 +1,158 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2018 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+// Class header
+#include "BufferUdp.h"
+
+// system headers
+
+
+// Third-party headers
+
+
+// qserv headers
+#include "loader/LoaderMsg.h"
+
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.loader.BufferUdp");
+}
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+/// Repeatedly read a socket until a valid MsgElement is read, eof,  or an error occurs.
+MsgElement::Ptr BufferUdp::readFromSocket(boost::asio::ip::tcp::socket& socket, std::string const& note) {
+    for (;;) {
+        LOGS(_log, LOG_LVL_DEBUG, note << " readFromSocket");
+        boost::system::error_code error;
+
+        // If there's something in the buffer already, get it and return.
+        // This can happen when the previous read of socket read multiple elements.
+        MsgElement::Ptr msgElem = _safeRetrieve();
+        if (msgElem != nullptr) {
+            return msgElem;
+        }
+
+        size_t len = socket.read_some(boost::asio::buffer(_wCursor, getAvailableWriteLength()), error);
+        _wCursor += len; /// must advance the cursor.
+
+        // EOF is only a problem if no MsgElement was retrieved.
+        // ??? This is definitely the case in UDP, EOF as nothing more will show up.
+        // ??? But TCP is another issue as EOF is returned when the connection is still live but
+        // ??? there's no data (len=0). Why does read_some set error to eof before the tcp connection is closed?
```

Comment:
yeah; according to the docs `eof` is because the peer closed the connection. You're sure the peer didn't close it?

Reply:
I'm pretty sure I've seen the warning below show up in the log, but it has been a little while since I went looking for it.